docs/pages/release_notes.md

```diff
@@ -13,13 +13,71 @@ This is a minor release.
 ### Table Of Contents
 
 * [New and noteworthy](#new-and-noteworthy)
+    *   [Drawing a line between private and public API](#drawing-a-line-between-private-and-public-api)
     *   [New Rules](#new-rules)
 * [Fixed Issues](#fixed-issues)
 * [API Changes](#api-changes)
 * [External Contributions](#external-contributions)
 
 ### New and noteworthy
 
+#### Drawing a line between private and public API
+
+Until now, all released public members and types were implicitly considered part
+of PMD's public API, including inheritance-specific members (protected members, abstract methods).
+We have maintained those APIs with the goal to preserve full binary compatibility between minor releases,
+only breaking those APIs infrequently, for major releases.
+
+
+In order to allow PMD to move forward at a faster pace, this implicit contract will
+be invalidated with PMD 7.0.0. We now introduce more fine-grained distinctions between
+the type of compatibility support we guarantee for our libraries, and ways to make
+them explicit to clients of PMD.
+
+##### `.internal` packages and `@InternalApi` annotation
+
+*Internal API* is meant for use *only* by the main PMD codebase. Internal types and methods
+may be modified in any way, or even removed, at any time.
+
+Any API in a package that contains an `.internal` segment is considered internal.
+The `@InternalApi` annotation will be used for APIs that have to live outside of
+these packages, e.g. methods of a public type that shouldn't be used outside of PMD (again,
+these can be removed anytime).
+
+##### `@ReservedSubclassing`
+
+Types marked with the `@ReservedSubclassing` annotation are only meant to be subclassed
+by classes within PMD. As such, we may add new abstract methods, or remove protected methods,
+at any time. All published public members remain supported. The annotation is *not* inherited, which
+means a reserved interface doesn't prevent its implementors to be subclassed.
+
+##### `@Beta`
+
+APIs marked with the `@Beta` annotation at the class or method level are subject to change.
+They can be modified in any way, or even removed, at any time. If your code is a library
+itself (i.e. it is used on the CLASSPATH of users outside your own control), you should not
+use beta APIs, unless you repackage them (e.g. using ProGuard, shading, etc).
+
+##### `@Experimental`
+
+
+APIs marked with the `@Experimental` annotation at the class or method level will almost certainly
+change. They can be modified in any way, or even removed, at any time. You should not use or rely
+ on them in any production code. They are purely to allow broad testing and feedback.
+
+##### `@Deprecated`
+
+APIs marked with the `@Deprecated` annotation at the class or method level will remain supported
+until the next major release but it is recommended to stop using them.
+
+
+##### The transition
+
+*All currently supported APIs will remain so until 7.0.0*. All APIs that are to be moved to
```

Comment:
This is correct. 

The one point I would be careful about is removing the annotation, I think it still may be useful in the future. For example, some methods on the AST need to stay public, even though they are still private API. I'm thinking for example of `setQualifiedName` on QualifiableNodes, which needs to remain public for the `qname` package to access. 

We could also consider making the `TypeNode` methods `setType`, `setTypeDefinition` and `getTypeDefinition` `@PrivateApi`, since we'll probably have some refactoring to do on `JavaTypeDefinition` to improve our type resolution, and the fact it's public API ties our hands for now... But of course the TypeNode interface shouldn't be moved to an internal package. `getTypeDefinition` could be made public API when it's stable. 

Overall I think the AST's API should be completely immutable for rule writers. Setters that we can't hide should be `@Deprecated` to raise a warning, and `@PrivateApi` provides finer disambiguation as to whether they're really to be removed than `forRemoval` 

Intuition tells me that the annotation could have other useful use cases, but that's only my two cents